src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs

```diff
@@ -192,13 +196,22 @@ public Task<TResult> ExecuteParallelForAllAsync<TResult>(IShardedOperation<TResu
             }
             catch (Exception e)
             {
-                if (typeof(TFailureMode) == typeof(ThrowOnFailure))
+                if (typeof(TFailureMode) == typeof(ThrowOnFirstFailure))
                     throw;
 
                 _exceptions ??= new Dictionary<int, Exception>();
                 _exceptions[shardNumber] = e;
             }
         }
+
+        if (typeof(TFailureMode) == typeof(ThrowAggregateFailure) && _exceptions?.Count > 0)
+        {
+            if (_exceptions.Count == 1)
+                throw _exceptions.First().Value;
+
+            throw new AggregateException($"Some shards ({string.Join(',', _exceptions.Keys)}) threw an exception", _exceptions.Values);
```

Comment:
Let's create a new exception here
`ExecutionForShardException : RavenException`
So you can specify which shard throw what

Reply:
Done